### ISSUE TITLE
Potential lack of keyword final

### ISSUE THREAD

**Issue by Xenoamor** [wovo/hwlib#23]
https://github.com/wovo/hwlib/blob/23d687944e7ce90c99c9c2b20f113cce3948c3f4/library/pins/hwlib-pin-all.hpp#L63

Looking at this line should the keyword final not be used? If this is committed then a vtable would need to be created and the generated code would bloat.

[I've made an example of this on godbolt](https://godbolt.org/z/S7DlGk)

**Comment by wovo** [comment]
As far as I can see the current version has 'override' on that line, maybe you refere to an older version?

The keyword override in itself doesn't add or remove the need for a vtable: that is cause by the 'virtual' in the supercalss (in this case whlib::pin_out).


Wouter van Ooijen

0638150444 - HL15 4.060

________________________________

**Comment by Xenoamor** [comment]
Hi Wouter, I think you might have misunderstood me. It is correct to have the keyword override but it would be beneficial to _also_ have the keyword final.
If you do then the compiler [can de-virtualise it](https://marcofoco.com/the-power-of-devirtualization/) and remove the overheads associated with it

**Comment by wovo** [comment]
OK, that makes sense. I didn't use final anywhere (yet), but I think there are no disadvantages. But In that case I should probably make the class final.

I'll give it a try (soon).


Wouter van Ooijen

0638150444 - HL15 4.060

________________________________

**Comment by Xenoamor** [comment]
Awesome, it'll be interesting to know if it becomes as fast as a templated version. I imagine link time optimisation might be needed for that though

**Comment by wovo** [comment]
I think even LTO won't optimize as agressively as a template-based version, especially not when multiple ports are used that are passed to multiple functions. IIRC a single port, passed to a single function, is indeed optimized. Tried that, a few years ago.

And with the templated version (godafoss) things are so much easier because not 'inbetween' objects are needed.


Wouter van Ooijen

0638150444 - HL15 4.060

________________________________

**Comment by Xenoamor** [comment]
Yes I see this now, I've played around with it in godbolt and you only get the savings from `final` if you are directly calling the final class. If you call anything above you get the intermediate objects as you say.

I'll have a look at godafoss when I get some free time!
